### PR TITLE
Add file imports

### PR DESCRIPTION
**Pull request by alexhumphreys**

### CODE REVIEW

File: Idrall/Resolve.idr [hunk 1]
```diff
@@ -3,75 +3,137 @@ module Idrall.Resolve
...
+
+liftEither : Either e a -> IOEither e a
+liftEither = MkIOEither . pure
+
+mapErr : (e -> e') -> IOEither e a -> IOEither e' a
+mapErr f (MkIOEither x) = MkIOEither (do
+  x' <- x
+  (case x' of
+        (Left l) => pure (Left (f l))
+        (Right r) => pure (Right r)))
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Resolve.idr:17]: These brackets look unnecessary. 
```suggestion
  case x' of
        (Left l) => pure (Left (f l))
        (Right r) => pure (Right r))
```

File: Idrall/Resolve.idr [hunk 2]
```diff
@@ -3,75 +3,137 @@ module Idrall.Resolve
...
+  resolveLocalFile : (history : List FilePath) -> (current : Maybe FilePath) -> (next : FilePath) -> IOEither Error (Expr Void)
+  resolveLocalFile h current next =
+    let combinedFilePaths = combinePaths current next in
+        go combinedFilePaths
+    where
+    go : FilePath -> IOEither Error (Expr Void)
+    go p = do
+      liftEither (alreadyImported h (normaliseFilePath p))
+      str <- readFile' (canonicalFilePath p)
+      expr <- mapErr (parseErrorHandler) (liftEither (parseExpr str))
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Resolve.idr:62]: Unnecessary brackets around `(parseErrorHandler)`.
```suggestion
      expr <- mapErr parseErrorHandler (liftEither (parseExpr str))
```

File: Idrall/Resolve.idr [hunk 3]
```diff
@@ -3,75 +3,137 @@ module Idrall.Resolve
...
+parseErrorHandler x = ErrorMessage (x)
+
+fileErrorHandler : String -> FileError -> Error
+fileErrorHandler x y = ReadFileError (show y ++ " " ++ x)
+
+readFile' : String -> IOEither Error String
+readFile' x =
+  let c = MkIOEither (readFile x)
+      contents = mapErr (fileErrorHandler x) c in
+  contents
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Resolve.idr:29]:
```suggestion
  let c = MkIOEither (readFile x) in
    contents = mapErr (fileErrorHandler x) c
```

File: Idrall/Check.idr [hunk 4]
```diff
@@ -203,6 +203,8 @@ mutual
   eval env (ENaturalIsZero x)
     = do x' <- eval env x
          doNaturalIsZero x'
+  eval env (EEmbed (Raw x)) = Left (ErrorMessage "Can't eval unresolved import")
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Check.idr:206]:
```suggestion
  eval env (EEmbed (Raw x)) = absurd x
```

File: Idrall/Check.idr [hunk 5]
```diff
@@ -526,3 +528,5 @@ mutual
     isList ctx xTy
     convert ctx (VConst CType) xTy yTy
     Right (xTy)
+  synth ctx (EEmbed (Raw _)) = Left (ErrorMessage "Can't synth unresolved import")
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Check.idr:531]:
```suggestion
  synth ctx (EEmbed (Raw x)) = absurd x
```

File: Idrall/Error.idr [hunk 6]
```diff
@@ -29,3 +31,5 @@ Show Error where
   show SortError = "SortError"
   show (AssertError str) = "AssertError" ++ str
   show (ListAppendError str) = "ListAppendError" ++ str
+  show (ReadFileError str) = "ReadFileError" ++ str
+  show (CyclicImportError str) = "CyclicImportError" ++ str
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Error.idr:35]:
```suggestion
  show (AssertError str) = "AssertError: " ++ str
  show (ListAppendError str) = "ListAppendError: " ++ str
  show (ReadFileError str) = "ReadFileError: " ++ str
  show (CyclicImportError str) = "CyclicImportError: " ++ str
```

File: Idrall/Resolve.idr [hunk 7]
```diff
@@ -3,75 +3,137 @@ module Idrall.Resolve
...
+      resolve (normaliseFilePath p :: h) (Just p) expr
+
   export
-  resolve : Expr ImportStatement -> IOEither Error (Expr Void)
-  resolve e@(EVar x) = pure e
-  resolve e@(EConst x) = pure e
-  resolve (EPi x y z) = do
-    y' <- resolve y
-    z' <- resolve z
+  resolve : (history : List FilePath) -> Maybe FilePath -> Expr ImportStatement -> IOEither Error (Expr Void)
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Resolve.idr:66]: Still thinking that this looks a lot like `traverse` :P

File: Idrall/Resolve.idr [hunk 8]
```diff
@@ -3,75 +3,137 @@ module Idrall.Resolve
...
-  resolveList (x :: xs) =
-    let rest = resolveList xs in
+  resolveList :  (history : List FilePath)
+              -> Maybe FilePath
+              -> List (Expr ImportStatement)
+              -> IOEither Error (List (Expr Void))
+  resolveList h p [] = MkIOEither (pure (Right []))
+  resolveList h p (x :: xs) =
+    let rest = resolveList h p xs in
     do rest' <- rest
```
Comment (symbiont-daniel-gustafsson) [inline Idrall/Resolve.idr:137]:
```suggestion
    do rest' <- resolveList h p xs
```
and I suppose you could remove the prime on `rest'`

File: Idrall/Check.idr (hunk 5 above)
Comment (alexhumphreys) [reply Idrall/Check.idr:531]: ooh absurd, cool

File: Idrall/Resolve.idr (hunk 7 above)
Comment (alexhumphreys) [reply Idrall/Resolve.idr:66]: Future me will make that happen somehow 😅

File: Idrall/Resolve.idr (hunk 3 above)
Comment (alexhumphreys) [reply Idrall/Resolve.idr:29]: think that's a syntax error, but i'll tidy it up in a similar way

**Pull request merged**